server/task.go

```diff
@@ -770,6 +770,29 @@ func (s *Server) patchTaskStatus(ctx context.Context, task *api.Task, taskStatus
 		}
 	}
 
+	// Cancel every dependency of the task.
+	if taskPatched.Status == api.TaskCanceled {
```

Comment:
Do we need a hard limit, e.g., 100, of the most canceled task to prevent any bug that introduces a circle in the DAG?

Reply:
Please extract the logic into another function to minimize the length.

Reply:
- Use map[int]bool to make sure we don't visit a node twice.
- Extracted.